src/core_ocean/shared/mpas_ocn_mesh.F

```diff
@@ -0,0 +1,1187 @@
+!|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||
+! \file mpas_ocn_mesh.F
+!
+! Copyright (c) 2013,  Los Alamos National Security, LLC (LANS)
+! and the University Corporation for Atmospheric Research (UCAR).
+!
+! Unless noted otherwise source code is licensed under the BSD license.
+! Additional copyright and license information can be found in the LICENSE file
+! distributed with this code, or at http://mpas-dev.github.com/license.html
+!
+!|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||
+!
+!  ocn_mesh
+!
+!>  \brief MPAS ocean mesh structure with GPU support
+!!  \author Rob Aulwes and Phil Jones
+!!  \date   14 Jan 2020
+!!  \details
+!!  This module creates and maintains a primary ocean mesh structure
+!!  and ensures all mesh variables are copied to an accelerator device
+!!  if needed. Currently it consists of pointers to the existing MPAS mesh pool
+!!  variables, but is intended to eventually replace the mesh pool later.
+!
+!-------------------------------------------------------------------------------
+
+module ocn_mesh
+
+   ! module dependencies
+   use mpas_derived_types
+   use mpas_pool_routines
+   use mpas_log
+
+   implicit none
+   private
+
+   !----------------------------------------------------------------------------
+   !
+   ! Public parameters
+   !
+   !----------------------------------------------------------------------------
+
+   integer, public :: &! mesh, array sizes
+      nCellsAll,      &! total number of local (owned+halo) cells in primary
```

Comment:
I really like the variables `nCellsAll, nCellsOwned, nCellsHalo(n)`. @philipwjones thanks for thinking of it. I myself can never remember which direction `nCellsArray` proceeds, and `nCells` sometimes has different definitions within the same module. As you use this `ocn_mesh` module in other modules, it will be good to switch from `nCells` to these exact specifications.

Reply:
Thanks and yeah, @mattdturner and I have been gradually replacing these as we start use-ing the module with the optimizations.